### ISSUE TITLE
G6 On new sensor start, data with timestamp before new sensor was started is sent to NS and AAPS

### ISSUE THREAD

**Issue by andyrozman** [NightscoutFoundation/xDrip#1890]
### Subject of the issue
When new G6 sensor is started, old data (data accumulated before new sensor was started) is being added to NS and send over broadcast. This seems like similar issue to https://github.com/NightscoutFoundation/xDrip/issues/1778 and solution might be the same. Once new sensor is started, all data from before should be ignored and dropped. Is this something that could be easily implemented? 

### Expected behavior
Data before sensor start is dropped.

### Actual behavior
When sensor starts all data is sent to NS and broadcast.

### Steps to reproduce the behavior:
Start new G6 sensor and look at data

### Additional notes
So new version of AAPS (3.0 - current betas), stops Loop as soon is it receives such data (new BgQualityPlugin). This is why all AAPS users are being warned not to use Xdrip anymore. If this could be fixed sooner, rather the later, warning about Xdrip probably won't go in release...

**Comment by Navid200** [comment]
When you attempted to open this issue, you saw a request that you consult with us before opening the issue.
Did you not see that note?

Would you please not use these issues as if this was a forum?

**Comment by Navid200** [comment]
As your questions like "is the solution the same as an existing issue", or "is this something that can be easily implemented" in the discussions:
https://github.com/Navid200/xDrip/discussions

I assure you having an open issue does not expedite resolving the issue.  It only clutters this platform. 
Please work with us so that we can efficiently maintain xDrip.  We need your help.
Please pay attention to the requests we make (notes you saw when you opened this issue).

**Comment by andyrozman** [comment]
I did ask on Gitter, but there was no response... It is true I didn't have all the details yet at that point, but this is valid issue, that will soon affect a lot of users of Xdrip (everybody that is using AndroidAPS that is).

**Comment by Navid200** [comment]
@andyrozman That may be the case.
But, if the cause is already included in an existing open issue, there is no reason to open another issue.
You can open a discussion and bring it to our attention.
We can label an existing issue and increase its priority if necessary.
But, having redundant issues doesn't help.

And if have a conversation, in discussions, and it turns out that the matter needs a dedicated issue, we can then open an issue.

We are just asking "please don't pull the trigger and ask questions later".
There are too many closed issues and people refuse to search them.  All because of so many unnecessary issues (some redundant, some just user error) that have been opened.

If you open a discussions, we can talk.

**Comment by savek-cc** [comment]
https://github.com/Navid200/xDrip/discussions/148 discusson for additional info on the topic

**Comment by savek-cc** [comment]
After looking at some of the source code, my guess is that xDrip decides to request a backfill (as the last bg values are quite old) on the first bg value from a new sensor, see https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/G5Model/Ob1G5StateMachine.java#L803 - will have to start a new sensor anyway in a couple of days; will see if I get around to having logcat open then at the 2h mark.

**Comment by ToreBj** [comment]
I can confirm that after a sensor change and warmup (2h) AAPS receives (a backfill) 3h of BG data. 

Why is a backfill requested in Xdrip?

**Comment by jamorham** [comment]
Is the problem that AAPS wont accept out of order data or that it wont accept data before the sensor start time?

**Comment by savek-cc** [comment]
The code in question sits in https://github.com/nightscout/AndroidAPS/blob/dev/app/src/main/java/info/nightscout/androidaps/plugins/constraints/bgQualityCheck/BgQualityCheckPlugin.kt#L79 - so it receives bg values from xDrip that are less than 20ms apart; so duplicate values with different timestamps.

**Comment by jamorham** [comment]
Isn't that 20 seconds expressed in ms? rather than 20ms difference?

**Comment by jamorham** [comment]
I'm pretty sure that xDrip is doing de-duplication of readings based on timestamp before it processes them. This should apply to upload to nightscout and local broadcast. There could be a bug with this code or something else could be happening. It would be very interesting if a user who has experienced this problem with AAPS could provide what the reading timestamps look like from that red warning code that @savek-cc linked - for example are they identical timestamps or slightly different?

**Comment by andyrozman** [comment]
@jamorham I think problem is that they are "slightly" different, but still within that 20s range. If they were the same NS and AAPS  would detect that, but they seem to be just little "off", so they get treated as new record.

**Comment by savek-cc** [comment]
@MilosKozak i thought that code was supposed to check for duplicates within a couple of ms - but apparently it really is 20 seconds that are checked here? Could you please provide a short explainaition from your understanding of the issue and what these 20 seconds are about?

**Comment by jamorham** [comment]
@andyrozman Is this problem for data which is being delivered from xDrip via local broadcast or after being uploaded to nightscout (or both) ?

**Comment by andyrozman** [comment]
I would imagine both. 
If you send it via broadcast there will be a problem and if it comes from NS also. AAPS look at all data it receives and checks for duplicates within those 20s...

**Comment by MilosKozak** [comment]
I'm not sure about NS and i don't know if xdrip even sends bg updates from the past to the NS.
 But it's definitely broadcasted to AAPS

**Comment by savek-cc** [comment]
@MilosKozak as I can't understand that check function: can you explain in plain English what that function checks for with the 20 seconds?

**Comment by ToreBj** [comment]
![Screenshot_20220105-141718](https://user-images.githubusercontent.com/23278852/149038116-75627a32-39d8-4afb-9d47-a98781d94c00.jpg)
Here you can see what is received from xdrip in AAPS .  You can see the 2h warm-up pause and the list was updated with duplicate BG values from before the sensor change, indicated with red after the first received BG. It seems that the full 3h transmitter buffer was received, i.e. almost an hour worth of data before sensor change had to be deleted for me to continue looping.

**Comment by ToreBj** [comment]
I had a look at the code that was referred, and I see that there is a calculation for the missing range, but can't see any testing for sensor warm-up?

**Comment by MilosKozak** [comment]
> @MilosKozak as I can't understand that check function: can you explain in plain English what that function checks for with the 20 seconds?

https://github.com/Navid200/xDrip/discussions/148

**Comment by MilosKozak** [comment]
> Isn't that 20 seconds expressed in ms? rather than 20ms difference?

received timestamps are in https://github.com/Navid200/xDrip/discussions/148

**Comment by savek-cc** [comment]
> > @MilosKozak as I can't understand that check function: can you explain in plain English what that function checks for with the 20 seconds?
> 
> [Navid200#148](https://github.com/Navid200/xDrip/discussions/148)

Ok - while that link still didn't explain the code to me, it just made "click" for me... 20s is still way to close for 2 values - I was somehow convinced that 20 seconds would be 4 readings; getting lost somewhere between ms, s and minutes there for whatever reason...

**Comment by savek-cc** [comment]
I just did a backup on my xDrip sqlite DB containing 50.000 entries and found this in there:
![Bildschirmfoto von 2022-01-12 21-48-51](https://user-images.githubusercontent.com/13501427/149219298-c193a265-66c6-47e0-856f-c335a2c84772.png)
First is OO Calc coloum number (ignore), then _id, caluclated_value, timestamp, diff(timestamp, prev timestamp)/1000 (so seconds)  - so most have a diff of ~300 seconds - but some are "0" or "1" second apart.
So as can be seen, there are backfill values with timestamps within 260-817ms of an original value within the actual xDip DB.
@jamorham

**Comment by savek-cc** [comment]
For anyone who wants to check on their own database (I used sqliteviewer on fedora):
`SELECT * from (
SELECT _id, calculated_value, timestamp, source_info,
(timestamp - LAG (timestamp, 1) OVER (ORDER BY timestamp)) as diff
FROM BgReadings WHERE source_info NOT NULL ORDER BY timestamp) WHERE diff < 20*1000;`
Will give you a table like this (which also the complete output of my instance)

_id | calculated_value | timestamp | source_info | diff
-- | -- | -- | -- | --
47916 | 139.0 | 1599799924438 | G5 Native | 635
47917 | 137.0 | 1599800224096 | G5 Native | 293
47918 | 134.0 | 1599800523975 | G5 Native | 172
47919 | 129.0 | 1599800823978 | G5 Native | 175
47920 | 128.0 | 1599801124137 | G5 Native | 334
47921 | 120.0 | 1599801424055 | G5 Native | 252
47922 | 121.0 | 1599801724117 | G5 Native | 314
47923 | 121.0 | 1599802023986 | G5 Native | 183
47924 | 118.0 | 1599802324022 | G5 Native | 219
47925 | 114.0 | 1599802623955 | G5 Native | 152
54450 | 103.0 | 1601796038765 | G5 Native | 466
54451 | 108.0 | 1601796338610 | G5 Native | 311
54452 | 89.0 | 1601796638767 | G5 Native | 468
54453 | 111.0 | 1601796939473 | G5 Native | 1174
54454 | 113.0 | 1601797238814 | G5 Native | 515
54455 | 115.0 | 1601797538611 | G5 Native | 312
54456 | 112.0 | 1601797838754 | G5 Native | 455
65259 | 86.0 | 1605074296270 | G5 Native | 5362
65260 | 84.0 | 1605074596328 | G5 Native | 5420
65261 | 82.0 | 1605074896647 | G5 Native | 5739
65262 | 81.0 | 1605075196757 | G5 Native | 6849
65263 | 64.0 | 1605075496628 | G5 Native | 6720
65264 | 62.0 | 1605075796649 | G5 Native | 6741
65265 | 75.0 | 1605076096475 | G5 Native | 6567
78329 | 156.0 | 1609092345784 | G5 Native | 271
78330 | 149.0 | 1609092645669 | G5 Native | 156
78331 | 146.0 | 1609092945629 | G5 Native | 116
78332 | 153.0 | 1609093245695 | G5 Native | 182
78333 | 148.0 | 1609093545651 | G5 Native | 138
78334 | 144.0 | 1609093845667 | G5 Native | 154
78335 | 134.0 | 1609094145581 | G5 Native | 68
78336 | 126.0 | 1609094445559 | G5 Native | 46
109398 | 76.0 | 1618555746780 | G5 Native::Backfill | 386
109399 | 79.0 | 1618556046780 | G5 Native::Backfill | 301
109400 | 80.0 | 1618556346780 | G5 Native::Backfill | 355
130372 | 76.0 | 1624908268654 | G5 Native | 961
130373 | 73.0 | 1624908569374 | G5 Native | 1681
130374 | 76.0 | 1624908869033 | G5 Native | 1340
141655 | 104.0 | 1628321231926 | G5 Native | 285
145675 | 75.0 | 1629530180350 | G5 Native | 817
145707 | 84.0 | 1629530479533 | G5 Native::Backfill | 372
145677 | 88.0 | 1629530779793 | G5 Native | 260
145709 | 87.0 | 1629531079533 | G5 Native::Backfill | 445
145710 | 90.0 | 1629531379533 | G5 Native::Backfill | 335
145711 | 72.0 | 1629531679533 | G5 Native::Backfill | 369

**Comment by savek-cc** [comment]
and when looking at the complete table at the timestamps in question, you can see that it apparently really is about the backfill after sensor change (the second to last one features a gap of ~129 minutes between the timestamps):

_id | calculated_value | timestamp | source_info | diff
-- | -- | -- | -- | --
47951 | 139.0 | 1599799923803 | G5 Native::Backfill | 292532
47916 | 139.0 | 1599799924438 | G5 Native | 635
47952 | 137.0 | 1599800223803 | G5 Native::Backfill | 299365
47917 | 137.0 | 1599800224096 | G5 Native | 293
47953 | 134.0 | 1599800523803 | G5 Native::Backfill | 299707
47918 | 134.0 | 1599800523975 | G5 Native | 172
47954 | 129.0 | 1599800823803 | G5 Native::Backfill | 299828
47919 | 129.0 | 1599800823978 | G5 Native | 175
47955 | 128.0 | 1599801123803 | G5 Native::Backfill | 299825
47920 | 128.0 | 1599801124137 | G5 Native | 334
47956 | 120.0 | 1599801423803 | G5 Native::Backfill | 299666
47921 | 120.0 | 1599801424055 | G5 Native | 252
47957 | 121.0 | 1599801723803 | G5 Native::Backfill | 299748
47922 | 121.0 | 1599801724117 | G5 Native | 314
47958 | 121.0 | 1599802023803 | G5 Native::Backfill | 299686
47923 | 121.0 | 1599802023986 | G5 Native | 183
47959 | 118.0 | 1599802323803 | G5 Native::Backfill | 299817
47924 | 118.0 | 1599802324022 | G5 Native | 219
47960 | 114.0 | 1599802623803 | G5 Native::Backfill | 299781
47925 | 114.0 | 1599802623955 | G5 Native | 152
47961 | 78.0 | 1599802923803 | G5 Native::Backfill | 299848
47950 | 187.0 | 1599810723604 | G5 Native | 7799801
47962 | 189.0 | 1599811025862 | G5 Native | 302258

**Comment by savek-cc** [comment]
Not all backfilled ones from before sensor change are bad though, check ID 54489 which is a valid backfill and does not cause a duplicate value. So the solution (as I guess, xDrip should not have these dupes in it's own database either @Navid200  @jamorham to actually check if the database already contains an entry at backfill-entry-timestamp +- xx seconds?)

_id | calculated_value | timestamp | source_info | diff_ms | time_str | sensor_age_min
-- | -- | -- | -- | -- | -- | --
54487 | 115.0 | 1601797538299 | G5 Native::Backfill | 299485 | 04-10-2020 07:45:38 | 123.3789
54455 | 115.0 | 1601797538611 | G5 Native | 312 | 04-10-2020 07:45:38 | 22400.5929666667
54488 | 112.0 | 1601797838299 | G5 Native::Backfill | 299688 | 04-10-2020 07:50:38 | 123.379033333333
54456 | 112.0 | 1601797838754 | G5 Native | 455 | 04-10-2020 07:50:38 | 22405.5953333333
54489 | 103.0 | 1601798138299 | G5 Native::Backfill | 299545 | 04-10-2020 07:55:38 | 123.379183333333
54481 | 189.0 | 1601806838103 | G5 Native | 8699804 | 04-10-2020 10:20:38 | 123.292883333333
54490 | 204.0 | 1601807138341 | G5 Native | 300238 | 04-10-2020 10:25:38 | 128.29685

**Comment by Navid200** [comment]
A Nightly was just released.

jamorham worked on this and it is meant to address the issues reported here.
Please report back when you get a chance to test it if the problem persists or not either way.
Thanks

**Comment by jamorham** [comment]
@savek-cc thanks for your analysis. Where you have close timestamps do these have a different sensor id column? I'm suspecting they do as the xDrip de-duplication is filtered by sensor session which maybe it also shouldn't be but it is possible for different sensor session id's to have duplicated backfilled data which I expect is likely part of the reason for this issue. What I have done is capped the backfill window start to the current sensor session start which I'm thinking will have a similar effect.

**Comment by savek-cc** [comment]
Hi @jamorham - here's the same data as in https://github.com/NightscoutFoundation/xDrip/issues/1890#issuecomment-1011466904 - with added cols:
_id | sensor | sensor_uuid | calculated_value | timestamp | source_info | diff_ms | time_str | sensor_age_min
-- | -- | -- | -- | -- | -- | -- | -- | --
47915 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 140.0 | 1599799631271 | G5 Native | 306677 | 11-09-2020 04:47:11 | 38730.7323666667
47951 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 139.0 | 1599799923803 | G5 Native::Backfill | 292532 | 11-09-2020 04:52:03 | 121.292533333333
47916 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 139.0 | 1599799924438 | G5 Native | 635 | 11-09-2020 04:52:04 | 38735.61845
47952 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 137.0 | 1599800223803 | G5 Native::Backfill | 299365 | 11-09-2020 04:57:03 | 121.292966666667
47917 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 137.0 | 1599800224096 | G5 Native | 293 | 11-09-2020 04:57:04 | 38740.6127833333
47953 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 134.0 | 1599800523803 | G5 Native::Backfill | 299707 | 11-09-2020 05:02:03 | 121.2932
47918 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 134.0 | 1599800523975 | G5 Native | 172 | 11-09-2020 05:02:03 | 38745.61075
47954 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 129.0 | 1599800823803 | G5 Native::Backfill | 299828 | 11-09-2020 05:07:03 | 121.293466666667
47919 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 129.0 | 1599800823978 | G5 Native | 175 | 11-09-2020 05:07:03 | 38750.6107666667
47955 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 128.0 | 1599801123803 | G5 Native::Backfill | 299825 | 11-09-2020 05:12:03 | 121.29365
47920 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 128.0 | 1599801124137 | G5 Native | 334 | 11-09-2020 05:12:04 | 38755.6134833333
47956 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 120.0 | 1599801423803 | G5 Native::Backfill | 299666 | 11-09-2020 05:17:03 | 121.2938
47921 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 120.0 | 1599801424055 | G5 Native | 252 | 11-09-2020 05:17:04 | 38760.6120833333
47957 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 121.0 | 1599801723803 | G5 Native::Backfill | 299748 | 11-09-2020 05:22:03 | 121.29395
47922 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 121.0 | 1599801724117 | G5 Native | 314 | 11-09-2020 05:22:04 | 38765.6131
47958 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 121.0 | 1599802023803 | G5 Native::Backfill | 299686 | 11-09-2020 05:27:03 | 121.294083333333
47923 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 121.0 | 1599802023986 | G5 Native | 183 | 11-09-2020 05:27:03 | 38770.6109166667
47959 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 118.0 | 1599802323803 | G5 Native::Backfill | 299817 | 11-09-2020 05:32:03 | 121.294233333333
47924 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 118.0 | 1599802324022 | G5 Native | 219 | 11-09-2020 05:32:04 | 38775.6115833333
47960 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 114.0 | 1599802623803 | G5 Native::Backfill | 299781 | 11-09-2020 05:37:03 | 121.294366666667
47925 | 4 | c264cf80-e396-44e8-b5c9-94c4ad20fb3e | 114.0 | 1599802623955 | G5 Native | 152 | 11-09-2020 05:37:03 | 38780.6103166667
47961 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 78.0 | 1599802923803 | G5 Native::Backfill | 299848 | 11-09-2020 05:42:03 | 121.2945
47950 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 187.0 | 1599810723604 | G5 Native | 7799801 | 11-09-2020 07:52:03 | 121.205866666667
47962 | 5 | 8dfe3337-8af3-4889-9833-a3e811af8a09 | 189.0 | 1599811025862 | G5 Native | 302258 | 11-09-2020 07:57:05 | 126.243516666667

**Comment by savek-cc** [comment]
@MilosKozak so https://github.com/NightscoutFoundation/xDrip/commit/6f7c3660d6e140d914f24a155a0bcffa810ab4d8 should fix the issue towards both NS and AAPS.
For reference: If you want to analyze and xDrip-Backup for this event, you can use the following SQL to get all pairs that are within 20 seconds of each other:
`SELECT * from (
SELECT _id, sensor, sensor_uuid, calculated_value, timestamp, source_info,
abs (timestamp - LAG (timestamp, 1) OVER (ORDER BY timestamp)) as diff_ms1,
abs (timestamp - LAG (timestamp, -1) OVER (ORDER BY timestamp)) as diff_ms2,
strftime('%d-%m-%Y %H:%M:%S', datetime(timestamp/1000, 'unixepoch')) time_str,
time_since_sensor_started / (1000*60) as sensor_age_min
FROM BgReadings WHERE source_info NOT NULL ORDER BY timestamp) where (diff_ms1 < 20000 or diff_ms2 < 20000);`

**Comment by ToreBj** [comment]
I have a sensor change tomorrow and will test this. Thank you!

**Comment by ToreBj** [comment]
I had a successful sensor change now. Well done👍

**Comment by ToreBj** [comment]
@savek-cc do you have some more details on how to connect and run SQL?

**Comment by Freloner** [comment]
> Is the problem that AAPS wont accept out of order data or that it wont accept data before the sensor start time?

Aaps accepts all data which creates duplicate thus loop malfunction.

**Comment by savek-cc** [comment]
> @savek-cc do you have some more details on how to connect and run SQL?

Use the import/export in xDrip to export the database. This will result in a zipped database backup on your phone. Copy that zip to your pc and extract it.
Afterwars you can use an SQLite browser to look at the data and run SQL statement on it. I used https://sqlitebrowser.org/ for that purpose.

**Comment by savek-cc** [comment]
@jamorham - do you plan on applying this fix also to the wear collector?
https://github.com/NightscoutFoundation/xDrip/blob/master/wear/src/main/java/com/eveningoutpost/dexdrip/G5Model/Ob1G5StateMachine.java#L764 and following

**Comment by andyrozman** [comment]
Was this fix already released? In which version?

**Comment by Navid200** [comment]
> Was this fix already released? In which version?

January 14, 2022.
https://navid200.github.io/xDrip/docs/Releases.html

**Comment by tolot27** [comment]
We cannot close this issue until it is fixed for wear as well. The code duplication requires changes twice.

**Comment by Navid200** [comment]
@andyrozman Thanks for reporting this.

Have you used the builds since January 14, 2022?  Is the issue addressed?
Can you use AAPS without problems now?

**Comment by savek-cc** [comment]
@Navid200 this is an issue that *sometimes* occured at sensor change - basically affecting anyone using xDrip with AndroidAPS.
Looking at the code changes and the provided samples, the issue should be resolved though. A final verdict (i.e. does this also happen on normal backfill or whatever) will probably only be possible after a considerate time interval (i.e. multiple month?) of observing the behaviour now.
As AAPS 3.0 was released this week, a lot more ppl would be affected now - so all AAPS 3 users should upgrade to an xDrip version that includes this fix.
